torch/optim/adagrad.py

```diff
@@ -103,13 +116,149 @@ def step(self, closure=None):
                     # record the step after step update
                     state_steps.append(state['step'])
 
-            F.adagrad(params_with_grad,
-                      grads,
-                      state_sums,
-                      state_steps,
-                      lr=group['lr'],
-                      weight_decay=group['weight_decay'],
-                      lr_decay=group['lr_decay'],
-                      eps=group['eps'])
+            adagrad(params_with_grad,
+                    grads,
+                    state_sums,
+                    state_steps,
+                    lr=group['lr'],
+                    weight_decay=group['weight_decay'],
+                    lr_decay=group['lr_decay'],
+                    eps=group['eps'],
+                    has_sparse_grad=has_sparse_grad,
+                    foreach=group['foreach'])
 
         return loss
+
+
+def adagrad(params: List[Tensor],
+            grads: List[Tensor],
+            state_sums: List[Tensor],
+            state_steps: List[int],
+            # kwonly args with defaults are not supported by functions compiled with torchscript issue #70627
+            # setting these as kwargs for now as functional API is compiled by torch/distributed/optim
+            has_sparse_grad: bool = False,
```

Comment:
Why is the default False here? (and not None)
Don't we want to check it if the user didn't provide it? Or raise an error if the user asked for foreach explicitly but didn't provide this?